src/NovelRT/Experimental/Graphics/Vulkan/VulkanGraphicsDevice.cpp

```diff
@@ -0,0 +1,524 @@
+        createInfo.imageArrayLayers = 1;
+        createInfo.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;
+
+        QueueFamilyIndices indices = Utilities::FindQueueFamilies(GetAdapter()->GetVulkanPhysicalDevice(), _surface);
+        uint32_t queueFamilyIndices[] = {indices.graphicsFamily.value(), indices.presentFamily.value()};
+
+        if (indices.graphicsFamily != indices.presentFamily)
+        {
+            createInfo.imageSharingMode = VK_SHARING_MODE_CONCURRENT;
+            createInfo.queueFamilyIndexCount = 2;
+            createInfo.pQueueFamilyIndices = queueFamilyIndices;
+        }
+        else
+        {
+            createInfo.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
+            createInfo.queueFamilyIndexCount = 0;
+            createInfo.pQueueFamilyIndices = nullptr;
+        }
+
+        createInfo.preTransform = swapChainSupport.capabilities.currentTransform;
+        createInfo.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
+        createInfo.presentMode = presentMode;
+        createInfo.clipped = VK_TRUE;
+        createInfo.oldSwapchain = VK_NULL_HANDLE;
+
+        VkResult swapChainResult = vkCreateSwapchainKHR(_device, &createInfo, nullptr, &_vulkanSwapchain);
+        if (swapChainResult != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException("Failed to create the VkSwapchainKHR.", swapChainResult);
+        }
+
+        _logger.logInfoLine("VkSwapchainKHR successfully created. Retrieving VkImages...");
+
+        VkResult imagesKHRQuery = vkGetSwapchainImagesKHR(_device, _vulkanSwapchain, &imageCount, nullptr);
+
+        if (imagesKHRQuery != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException("Failed to retrieve the VkImages from the VkSwapchainKHR.",
+                                                             imagesKHRQuery);
+        }
+
+        _swapChainImages.resize(imageCount);
+        imagesKHRQuery = vkGetSwapchainImagesKHR(_device, _vulkanSwapchain, &imageCount, _swapChainImages.data());
+
+        if (imagesKHRQuery != VK_SUCCESS)
+        {
+            throw Exceptions::InitialisationFailureException("Failed to retrieve the VkImages from the VkSwapchainKHR.",
+                                                             imagesKHRQuery);
+        }
+
+        _vulkanSwapChainFormat = surfaceFormat.format;
+        _swapChainExtent = extent;
+
+        _logger.logInfoLine("VkImages successfully retrieved.");
+    }
+
+    void VulkanGraphicsDevice::Initialise()
+    {
+        CreateLogicalDevice();
+        CreateSwapChain();
+
+        _logger.logInfoLine("Vulkan logical device version 1.2 has been successfully initialised.");
+    }
+
+    void VulkanGraphicsDevice::TearDown()
+    {
+        vkDestroySwapchainKHR(_device, _vulkanSwapchain, nullptr);
+        vkDestroyDevice(_device, nullptr);
+
+        _logger.logInfoLine("Vulkan logical device version 1.2 successfully torn down.");
+    }
```

Comment:
Doesn't look this this is currently called via the sample application. Is that correct?

Reply:
I believe this is called by the dtor. It's built this way so the device can be "recycled" as well as completely deleted safely.